### ISSUE TITLE
Should S4018 flag a violation when a method returns <T>?

### ISSUE THREAD

**Issue by davejclausen** [SonarSource/sonarlint-visualstudio#255]
Hi, S4018 is flagging a violation on this method:

```c#
public T Deserialize<T>(string value) 
```

Is that correct/expected behavior?

**Comment by valhristov** [comment]
Hi @davejclausen, S4018 is supposed to enforce writing method signatures that help the compiler to imply the generic arguments from the types of the method parameters, e.g. a generic argument should not be used only as return type. Of course, using the generic argument as return type does not lead to bugs, just when using the method you need to write a little more explicit code. Perhaps we should add hardcoded exclusion for methods named "deserialize", but for the time being I would recommend just marking the issue as false positive.

**Comment by davejclausen** [comment]
Ok, no problem.  I was thinking that perhaps the rule could exempt methods that return <T>...

**Comment by Evangelink** [comment]
@valhristov If you think we can do anything please reopen an issue in SonarC# GitHub and close this one.

**Comment by Evangelink** [comment]
Hi @davejclausen,
We decided to lower the severity of the rule but to keep reporting on all cases as the number of methods considered as FP is pretty low. Besides with the upcoming feature of SonarLint for VS (don't show suppressed errors in connected mode) we think it is better to mark the issue as FP on SonarQube.

Cheers,
Amaury

**Comment by JohnyWS** [comment]
Perhaps using the convetion of System.Func with naming the generic parameter "TReturn" would be a better alternative? I'd even argue that having that excemption in place would make it fine to increse the severity of the method again. :)
